src/UnityExtension/Assets/Editor/GitHub.Unity/UI/TreeControl.cs

```diff
@@ -0,0 +1,465 @@
+                        }
+                    }
+                }
+            }
+            foldersKeys = Folders.Keys.Cast<string>().ToList();
+        }
+
+        public Rect Render(Rect rect, Vector2 scroll, Action<TreeNode> singleClick = null, Action<TreeNode> doubleClick = null)
+        {
+            Profiler.BeginSample("TreeControl");
+            bool visible = true;
+            var availableHeight = rect.y + rect.height;
+
+            RequiresRepaint = false;
+            rect = new Rect(0f, rect.y, rect.width, ItemHeight);
+
+            var titleNode = nodes[0];
+            ResetNodeIcons(titleNode);
+            bool selectionChanged = titleNode.Render(rect, 0f, selectedNode == titleNode, FolderStyle, TreeNodeStyle, ActiveTreeNodeStyle);
+
+            if (selectionChanged)
+            {
+                ToggleNodeVisibility(0, titleNode);
+            }
+
+            RequiresRepaint = HandleInput(rect, titleNode, 0);
+            rect.y += ItemHeight + ItemSpacing;
+
+            Indent();
+
+            int level = 1;
+            int i = 1;
+            for (; i < nodes.Count; i++)
+            {
+                var node = nodes[i];
+                ResetNodeIcons(node);
+
+                if (node.Level > level && !node.IsHidden)
+                {
+                    Indent();
+                }
+
+                if (visible)
+                {
+                    var changed = node.Render(rect, Styles.TreeIndentation, selectedNode == node, FolderStyle, TreeNodeStyle, ActiveTreeNodeStyle);
+
+                    if (node.IsFolder && changed)
+                    {
+                        // toggle visibility for all the nodes under this one
+                        ToggleNodeVisibility(i, node);
+                    }
+                }
+
+                if (node.Level < level)
+                {
+                    for (; node.Level > level && indents.Count > 1; level--)
+                    {
+                        Unindent();
+                    }
+                }
+                level = node.Level;
+
+                if (!node.IsHidden)
+                {
+                    if (visible)
+                    {
+                        RequiresRepaint = HandleInput(rect, node, i, singleClick, doubleClick);
+                    }
+                    rect.y += ItemHeight + ItemSpacing;
+                }
+            }
+
+            Unindent();
+
+            Profiler.EndSample();
+            return rect;
+        }
+
+        public void Focus()
+        {
+            bool selectionChanged = false;
+            if (Event.current.type == EventType.KeyDown)
+            {
+                int directionY = Event.current.keyCode == KeyCode.UpArrow ? -1 : Event.current.keyCode == KeyCode.DownArrow ? 1 : 0;
+                int directionX = Event.current.keyCode == KeyCode.LeftArrow ? -1 : Event.current.keyCode == KeyCode.RightArrow ? 1 : 0;
+                if (directionY != 0 || directionX != 0)
```

Comment:
I'm a bit confused by this set of if statements. I don't fully understand the functionality here, so that just might be my confusion, but it seems like a set of nested ifs with two different variables shouldn't just be an or?

Reply:
You are right, the wrapping if statement is redundant.

The local and remote trees are two separate tree controls.
This code lets you travel from one tree control to the other via keystroke if the scenario is correct.